prov/rxm/src/rxm_ep.c

```diff
@@ -936,16 +958,238 @@ void rxm_ep_progress_deferred_queue(struct rxm_ep *rxm_ep,
+
+	ret = fi_allreduce(coll_ep, buf, count, desc, result, result_desc,
+			   coll_addr, datatype, op, flags, req);
+	if (ret) {
+		ofi_ep_lock_acquire(&rxm_ep->util_ep);
+		rxm_free_coll_buf(rxm_ep, req);
+		ofi_ep_lock_release(&rxm_ep->util_ep);
+	}
+	return ret;
+}
+
+ssize_t rxm_ep_allgather(struct fid_ep *ep, const void *buf, size_t count,
+			 void *desc, void *result, void *result_desc,
+			 fi_addr_t coll_addr, enum fi_datatype datatype,
+			 uint64_t flags, void *context)
+{
+	struct rxm_ep *rxm_ep;
+	struct fid_ep *coll_ep;
+	struct rxm_coll_buf *req;
+	ssize_t ret;
+
+        rxm_ep = container_of(ep, struct rxm_ep, util_ep.ep_fid.fid);
+
+        ofi_ep_lock_acquire(&rxm_ep->util_ep);
+	req = rxm_get_coll_buf(rxm_ep);
+        ofi_ep_lock_release(&rxm_ep->util_ep);
+
+	if (!req)
+		return -FI_EAGAIN;
+
+	req->ep = rxm_ep;
+	req->flags = flags;
+	req->app_context = context;
+
+	coll_ep = get_coll_ep(rxm_ep, flags, FI_ALLGATHER);
+	flags &= ~FI_PEER_TRANSFER;
+
+	ret = fi_allgather(coll_ep, buf, count, desc, result, result_desc,
+			   coll_addr, datatype, flags, req);
+	if (ret) {
+		ofi_ep_lock_acquire(&rxm_ep->util_ep);
+		rxm_free_coll_buf(rxm_ep, req);
+		ofi_ep_lock_release(&rxm_ep->util_ep);
+	}
+	return ret;
+}
+
+ssize_t rxm_ep_scatter(struct fid_ep *ep, const void *buf, size_t count,
+		       void *desc, void *result, void *result_desc,
+		       fi_addr_t coll_addr, fi_addr_t root_addr,
+		       enum fi_datatype datatype, uint64_t flags,
+		       void *context)
+{
+	struct rxm_ep *rxm_ep;
+	struct fid_ep *coll_ep;
+	struct rxm_coll_buf *req;
+	ssize_t ret;
+
+        rxm_ep = container_of(ep, struct rxm_ep, util_ep.ep_fid.fid);
+
+        ofi_ep_lock_acquire(&rxm_ep->util_ep);
+	req = rxm_get_coll_buf(rxm_ep);
+        ofi_ep_lock_release(&rxm_ep->util_ep);
+
+	if (!req)
+		return -FI_EAGAIN;
+
+	req->ep = rxm_ep;
+	req->flags = flags;
+	req->app_context = context;
+
+	coll_ep = get_coll_ep(rxm_ep, flags, FI_SCATTER);
+	flags &= ~FI_PEER_TRANSFER;
+
+	ret = fi_scatter(coll_ep, buf, count, desc, result, result_desc,
+			 coll_addr, root_addr, datatype, flags, req);
+	if (ret) {
+		ofi_ep_lock_acquire(&rxm_ep->util_ep);
+		rxm_free_coll_buf(rxm_ep, req);
+		ofi_ep_lock_release(&rxm_ep->util_ep);
+	}
+	return ret;
+}
+
+ssize_t rxm_ep_broadcast(struct fid_ep *ep, void *buf, size_t count,
+			 void *desc, fi_addr_t coll_addr, fi_addr_t root_addr,
+			 enum fi_datatype datatype, uint64_t flags,
+			 void *context)
+{
+	struct rxm_ep *rxm_ep;
+	struct fid_ep *coll_ep;
+	struct rxm_coll_buf *req;
+	ssize_t ret;
+
+        rxm_ep = container_of(ep, struct rxm_ep, util_ep.ep_fid.fid);
+
+        ofi_ep_lock_acquire(&rxm_ep->util_ep);
+	req = rxm_get_coll_buf(rxm_ep);
+        ofi_ep_lock_release(&rxm_ep->util_ep);
+
+	if (!req)
+		return -FI_EAGAIN;
+
+	req->ep = rxm_ep;
+	req->flags = flags;
+	req->app_context = context;
+
+	coll_ep = get_coll_ep(rxm_ep, flags, FI_BROADCAST);
+	flags &= ~FI_PEER_TRANSFER;
```

Comment:
There's a lot of duplicate code in the above collective calls.  I think the only change is the collective enum value (e.g. FI_BROADCAST)

Reply:
Let me see if they can be replaced with a common routine.